### ISSUE TITLE
État original du projet et travaux envisagés

### ISSUE THREAD

**Issue by slithiaote** [victorgalmiche/umap-clustering-mise-en-prod#2]
# État original du projet
Tel qu’il est au moment où vous le reprenez pour ce cours (modélisation, état des bonnes pratiques, déploiement éventuel etc). L’idée est de pouvoir comprendre de quelle base vous partez et de vous pouvoir évaluer la progression du projet plutôt que simplement l’état final.

# Brève description du projet original et de son objectif
[`umap-clustering`](https://github.com/paquitopg/umap-clustering) est un travail de groupe réalisé par Paco Goze, Victor Galmiche et Matéo Morin dans le cadre du cours d'Advanced Machine Learning de 3e année de l'ENSAE. C'est un mini projet de recherche avec pour objectif de comprendre, d'implémenter et d'évaluer l'algorithme UMAP avec du clustering en couche additionnelle.

Les points importants du projet initial sous git-hub sont :
- Un algorithme UMAP (`umap_algo/`) fait main en suivant [le papier de recherche original](https://arxiv.org/abs/1802.03426) : même s'il n'arrive pas à gérer des grandes quantités de données (problème de recherche de valeurs singulières), il peut montrer la construction des embeddings en direct (gradient descent).
- Des scripts d'évaluation (`umap_comparisons/`) employant différentes méthodes de clustering sur un panel de données auxquelles l'algorithme UMAP de la librairie `umap-learn` a été appliqué.
- Un dossier `tests/`obsolète, retraçant des premiers essais mais qui n'a jamais été supprimé.

## Bonnes pratiques déjà mises en place

- [x] **Utilisation de `Git`**
    * Présence d'un fichier `.gitignore` adapté au langage et avec des règles additionnelles pour respecter les bonnes pratiques de _versioning_
    * Travail collaboratif : utilisation des branches et des *pull requests*
- [x]  **Présence d'un fichier `README`** présentant le projet : contexte, objectif, comment l'utiliser ?
    * Un `README` qui déclare déjà bien la structure du projet et offre un script d'exemple &rarr; à améliorer dans #3 
- [x]  **Présence d'un fichier `LICENSE`** déclarant la licence (_open-source_) d'exploitation du projet.
- [ ]  **Versioning des packages** : présence d'un fichier `requirements.txt` ou d'un fichier d'environnement `environment.yml` pour `conda`
    * `requirements.txt` est fourni, mais `pip install -r requirements.txt` impose à l'utilisateur d'installer globalement les librairies nécessaires &rarr; passer à `uv` #1 
- [ ]  **Qualité du code**
    + Respect des standards communautaires : utiliser un _linter_ et/ou un _formatter_
    + Modularité : un script principal qui appelle des modules
    + Partie code `umap_algo/` commentée et structurée &rarr; revoir la mise en page (Ruff) et se pose la question de la nécessité d'une classe
- [ ] **Structure des projets**
    + Respect des standards communautaires (`cookiecutter`)
    + Modularité du projet selon le modèle évoqué dans le cours:
        * Code sur `GitHub`
        * Données sur `S3`
        * Fichiers de configuration (_secrets_, etc.) à part

# Travaux prévus de mise en production
L'algorithme UMAP fait-main étant limité, il ne sera utilisé que pour des objectifs d'animation (site web). Pour le reste, on s'en référera à la bibliothèque `umap-learn` qui propose un `fit_transform()` et un `transform()` (qui n'est pas vraiment le `transform()` que l'on imagine pour du Machine Learning habituel, et qui est donc à manipuler avec attention).

## 1. API cf #16 
- Mettre en place une API qui `fit_transform()` les données (attention au volume à traiter car les délais vont exploser)
- Paramétriser correctement l'algorithme avec des fichiers de configuration > #11 

## 2. Site web / front end pour l'API cf #18 
- Utilisation de Streamlit
- Jeux de données déjà fournis
- Animation UMAP
- Application de clustering
- Visualisation des résultats finaux

## 4. Mettre en place les outils pour du CI/CD
- Mettre en place des tests
- Github Actions avec les tests pour chaque nouvelle version
- Argo CD pour actualiser les serveurs hébergeant le Streamlit et l'API

## 5. Autres
- Ajouter un script d'exemple dans le répertoire test pour faciliter le premier test du code par un nouvel utilisateur
- run `ruff check`
- Ajouter un "guide du développeur"

# Remarques

## Pourquoi pas un package ?
- l'état de départ est déjà très proche d'un package, peu de travail à effectuer
- moins dans l'esprit du cours

## Exemples d'utilisation
